### ISSUE TITLE
RFC/POC: end-to-end test cases as ansible plays

### ISSUE THREAD

**Issue by jrusz** [osbuild/osbuild-composer#655]
I want to propose creating our end-to-end test scenarios as ansible plays that could be ran with schutzbot. I think this will mimic how a user would use osbuild-composer which is what we want. What I'm not sure about is if this is the way we want to implement it. My first play looks like this:
```
- name: "first test"
  hosts: localhost
  tasks:
    - name: "Make sure osbuild-composer.socket is installed"
      dnf:
        name:
          - osbuild-composer
          - composer-cli
        state: present
    - name: "Make sure osbuild-composer is started"
      service:
        name: osbuild-composer.socket
        state: started
    - name: "Create a basic blueprint"
      copy:
        dest: "base.toml"
        content: |
          name = "base"
          description = "A base system with bash"
          version = "0.0.1"
          modules = []
          groups = []
 
          [[packages]]
          name = "bash"
          version = "*"
 
          [customizations]
          hostname = "test-image"
 
          [[customizations.user]]
          name = "test_user"
          password = "$6$d4C/rAdVMPPoCCoq$a20HKMJV406hbwHmazr6xWzvYYbUyDCrJDl.vFysg0ZdHW282JfckqtEsPLdQnzWNy2b6zj26cQVJPq/gd8HT/"
          groups = ["wheel"]
    - name: "Upload the blueprint"
      command: "composer-cli blueprints push base.toml"
    - name: "Start an qcow2 compose"
      command: "composer-cli compose start base qcow2"
      register: result
    - set_fact:
        UUID: "{{ result.stdout | regex_search('\\S{36}') }}"
    - name: "Wait for compose to finish"
      command: "composer-cli compose status"
      register: result
      delay: 15
      retries: 60
      until: result.stdout.find("FINISHED") != -1
    - name: "Download finished image"
      command: "composer-cli compose image {{ UUID }}"
    - name: "start libvirtd"
      service:
        name: libvirtd.service
        state: started
    - name: "create vm"
      command: "virt-install -n foobar --memory 2048 --nographics --import --disk {{ UUID }}-disk.qcow2 --noautoconsole --os-variant fedora32"
    - name: "check VM"
      virt:
        name: foobar
```
I suppose that this would share a common setup phase and possibly use some clean up phase. I'm not sure about the end if this check is enough, maybe a successful ssh connection into the booted VM would be nice to have. @major already tried something like this over at ansible-osbuild https://github.com/osbuild/ansible-osbuild/blob/master/schutzbot/composer-smoke-test.yml but we agreed it should be in osbuild-composer repo. Maybe have another task in https://github.com/osbuild/osbuild-composer/blob/master/schutzbot/test.yml and include these kind of scenarios there similar to how the other tests are ran now? Any thoughts about this @atodorov @teg ?

**Comment by atodorov** [comment]
> I want to propose creating our end-to-end test scenarios as ansible plays

I am not going to comment on the technical approach ATM b/c almost any approach would work. 

We should first discuss if/how we want e2e testing, the extent of it, how much does it overlap with the existing test suite (both in this repo and in lorax's repo), etc. Technical implementation is a secondary thing.

> I think this will mimic how a user would use osbuild-composer which is what we want

To be fair probably the majority of users will be using Cockpit's UI instead of the cli. I don't have the numbers but probably Terry and Jenny do have insights here.

As a technical team we want to be confident that the software works as it is expected, which doesn't necessarily translate 1:1 how the user is going to interact with it. 

FTR in the "Test Pyramid" e2e testing sits at the top and is there for a reason. It is usually slower, more brittle and harder to implement than testing at other levels of the stack. It does have its purpose but in virtually every subject on testing you will read that this should be avoided unless you can't perform testing in any other way.  That is why you can see testing activities designed in levels - unit, component (e.g. API in our case), integration, UI, e2e - that is the preference is always to try doing more testing at the bottom of the pyramid.


@jrusz what is the motivation behind requesting e2e testing, other than "that's how the user is going to use it" ? Is there a scenario/area which you have identified which isn't covered or poorly covered by the existing test suites ?  IMO we need to have specific reasons for introducing more variation into the current testing stack and everyone should feel these reasons are justified.

**Comment by teg** [comment]
Could we achieve this by extending our current composer-cli tests? I have no objection to adding some end to end test (if it does not slow things down too much), but isn't it already sort of what we do with composer-cli? My main concern is that I would like the actual tests to live in the tests subpackage, and let the ansible playbook simply be responsible for running them.

**Comment by jrusz** [comment]
@atodorov Thanks for this explanation. My further motivation behind this is that issue like #524 could be discovered with this kind of test. Whenever I tried this kind of test scenario for the past few releases I got some kind of error.. Maybe this could be covered by extending those CLI test but I was thinking that it would bring value to have at least a few test cases where it would test this as a whole and not in levels like you described. But I don't have as much experience so that's why I'm asking.. I'll do some more research and see if I can find something else this would help covering.

**Comment by atodorov** [comment]
I have commented about #524 separately. 

> Whenever I tried this kind of test scenario for the past few releases I got some kind of error..

That's very valuable info b/c we seem to be having at least some tests in this area and first I am interested to know why didn't they fail and if extending them where needed would cause a failure, e.g. this issue to be detected.  Let's figure out the root cause first and then propose solutions not the other way around.

**Comment by jikortus** [comment]
As a follow-up to this discussion and to one end-to-end scenario @major implemented as a [smoke test](https://github.com/osbuild/ansible-osbuild/blob/master/schutzbot/composer-smoke-test.yml), I tried to think of some more scenarios we could test. Not that I could come with anything really extraordinary and possibly some of the following might not be entirely relevant or could be covered by the CLI tests, but here are a few thoughts on the topic:

At this point, it seems most of the functionality that could be tested is or will be covered by the existing test-suite on different levels. Still, I think e2e tests could be useful, as they would make the testing a bit closer to real-world use that e. g. the existing image tests. However, they shouldn't be numerous, at least until some more knowledge about real-world users' scenarios has been obtained.

Some possible scenarios we could test:
- starting and cancelling a compose, and then starting a new one that should finish successfully (not implemented ATM, see PR #180)
- setting up a cloud provider profile from CLI and starting a compose+upload (this and the following item is blocked ATM by missing support for providers and upload handling via CLI)
- setting up a cloud provider and uploading an existing image
- cancelling a running upload to cloud (and retrying again)
- possibly we could introduce some scenarios leading to an expected crash, and test a subsequent recovery (e. g. compose crashed when running out of disk space), or (although this would/should not be a common scenario and probably should be defined based on some concrete expectations) testing osbuild-composer's ability to start again after it crashed for some reason

**Comment by atodorov** [comment]
We haven't seen the need for an e2e test suite for a long time now so closing.
